File: exercise/scala/day02/src/test/scala/FizzBuzzProperties.scala
import games.FizzBuzz
import org.scalacheck.Gen
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.scalacheck.ScalaCheckPropertyChecks

import scala.runtime.stdLibPatches.Predef.assert

class FizzBuzzProperties extends AnyFunSuite with ScalaCheckPropertyChecks {
  val MIN: Int = FizzBuzz.MIN
  val MAX: Int = FizzBuzz.MAX

  val fizzBuzzStrings: List[String] = List("Fizz", "Buzz", "FizzBuzz")

  def validStringsFor(x: Int): List[String] = fizzBuzzStrings :+ x.toString

  test("parse return a valid string for numbers between 1 and 100") {
    forAll(Gen.chooseNum(MIN, MAX)) { x =>
      assert(
        FizzBuzz.convert(x)
          .exists(result => validStringsFor(x).contains(result))
      )
    }
  }

  test("parse fail for numbers out of range") {
    forAll(Gen.chooseNum(Int.MinValue, Int.MaxValue).suchThat(i => i < FizzBuzz.MIN || i > FizzBuzz.MAX)) { x =>
      assert(FizzBuzz.convert(x).isEmpty)
    }
  }
}

File: solution/day02/src/test/kotlin/FizzBuzzProperties.kt
import games.FizzBuzz
import games.MAX
import games.MIN
import io.kotest.core.spec.style.StringSpec
import io.kotest.property.Arb
import io.kotest.property.arbitrary.filter
import io.kotest.property.arbitrary.int
import io.kotest.property.forAll

fun Map<Int, String>.generateValidStrings(): List<String> =
    this.values + flatMap { first ->
        filter { second -> second.key > first.key }
            .map { second -> first.key * second.key to first.value + second.value }
    }.toMap().values

fun validStringsFor(x: Int): List<String> = Configuration.mapping.generateValidStrings() + x.toString()
val validNumbers = Arb.int(MIN..MAX)

class FizzBuzzProperties : StringSpec({

    "parse return a valid string for numbers between 1 and 100" {
        forAll(validNumbers) { x ->
            FizzBuzz.convert(Configuration.mapping, x).isSome { result -> validStringsFor(x).contains(result) }
        }
    }

    "parse fail for numbers out of range" {
        forAll(Arb.int().filter { i -> i < MIN || i > MAX }) { x ->
            FizzBuzz.convert(Configuration.mapping, x).isNone()
        }
    }

    "parse fail for valid numbers with no configuration" {
        forAll(validNumbers) { x ->
            FizzBuzz.convert(emptyMap(), x).isNone()
        }
    }
})